modules/functional_expansion_tools/src/utils/MutableCoefficientsInterface.C

```diff
@@ -0,0 +1,220 @@
+{
+  // Nothing here
+}
+
+MutableCoefficientsInterface::MutableCoefficientsInterface(
+    const InputParameters & parameters,
+    const std::vector<Real> & new_coefficients,
+    bool enforce_size)
+  : _characteristics(0),
+    _coefficients(new_coefficients),
+    _enforce_size(enforce_size),
+    _print_coefficients(parameters.get<bool>("print_when_set"))
+{
+  // Nothing here
+}
+
+MutableCoefficientsInterface::MutableCoefficientsInterface(const InputParameters & parameters,
+                                                           std::vector<Real> && dropin_coefficients,
+                                                           bool enforce_size)
+  : _characteristics(0),
+    _coefficients(dropin_coefficients),
+    _enforce_size(enforce_size),
+    _print_coefficients(parameters.get<bool>("print_when_set"))
+{
+  // Nothing here
+}
+
+MutableCoefficientsInterface::MutableCoefficientsInterface(const InputParameters & parameters,
+                                                           std::size_t size,
+                                                           Real fill,
+                                                           bool enforce_size)
+  : _characteristics(0),
+    _coefficients(size, fill),
+    _enforce_size(enforce_size),
+    _print_coefficients(parameters.get<bool>("print_when_set"))
+{
+  // Nothing here
+}
+
+Real MutableCoefficientsInterface::operator[](std::size_t index) const
+{
+  return _coefficients[index];
+}
+
+const std::vector<std::size_t> &
+MutableCoefficientsInterface::getCharacteristics() const
+{
+  return _characteristics;
+}
+
+const std::vector<Real> &
+MutableCoefficientsInterface::getCoefficients() const
+{
+  return _coefficients;
+}
+
+std::size_t
+MutableCoefficientsInterface::getSize() const
+{
+  return _coefficients.size();
+}
+
+bool
+MutableCoefficientsInterface::isCompatibleWith(const MutableCoefficientsInterface & other) const
+{
+  // Check the coefficient sizes if requested
+  if ((_enforce_size && other._enforce_size) && getSize() != other.getSize())
+    return false;
+
+  // Check the size of the characteristics array
+  if (_characteristics.size() != other._characteristics.size())
+    return false;
+
+  // Check the values of the characteristics array
+  for (std::size_t i = 0; i < _characteristics.size(); ++i)
+    if (_characteristics[i] != other._characteristics[i])
+      return false;
+
+  return true;
+}
+
+bool
+MutableCoefficientsInterface::isSizeEnforced() const
+{
+  return _enforce_size;
+}
+
+void
+MutableCoefficientsInterface::enforceSize(bool enforce)
+{
+  _enforce_size = enforce;
+}
+
+void
+MutableCoefficientsInterface::importCoefficients(const MutableCoefficientsInterface & other)
+{
+  if (!isCompatibleWith(other))
+    mooseError("Cannot import coefficients from incompatible MutableCoefficientsInterface");
+
+  _coefficients = other._coefficients;
+
+  if (_print_coefficients)
+    dynamic_cast<ConsoleStreamInterface *>(this)->_console << *this;
```

Comment:
This looks like an instant segfault. `MutableCoefficientsInterface` does not inherit from `ConsoleStreamInterface`, so the result of you cast will be `nullptr` and doing `->` on it is bad.

Reply:
All the classes that use this interface (Function and UserObject, currently) do inherit from `ConsoleStreamInterface`. It is only intended to be used with these, and I am avoiding a diamond-inheritance scheme.

If working code is a concern, I can verify that it does work. I created the test `multiapp_print_coefficients.i` to ensure this works in the future as well.

If there is a design issue with this then I will gladly work with the team to come up with an alternative.

Reply:
Also, I am not convinced about the `_print_coefficient` thing.  I'd expect you provide a getter, so people can get the coefficients and then they can print them using the regular `_console << data` syntax.

Reply:
I have no problem providing a `Stream` object containing a formatted table of the coefficients. Is this what you are referring to?

I already provide access to the raw data as a vector via the `GetCoefficients()` method. The input option `print_coefficients` is available for the users who are using the as-built module, not developing their own extensions.

Reply:
A getter `std::string getCoefficientsTable()` was just added.

Reply:
There are multiple ways how to go about printing. I was suggesting that this class would only provide the raw data and the printing would be done by the caller. But, then I saw you overloaded `operator <<` and have some pretty printing there. So, I think you probably have a good reason to do so. My guess is that you do are reusing this class and do not want to replicate the code.

Providing a `std::string` through a function would be another good way.

Another solution to this problem is to pass the pointer to `MooseObject`-derived class that is using this interface and use that to get `_console` member, see `SetupInterface` for an example of what I am talking about.  Actually, that should be probably the way you should pick, so that your interface is consistent with the other interfaces in the framework.

Reply:
Exactly, this class is reused a few times. I'll keep the `std::string` from the function as it is very small to implement. I do like the `SetupInterface` approach, so I'll transition the casting over.

Reply:
Done. I also moved `MutableCoefficientsInterface` from utils/ to a directory that better represents its assignment in the context of functional expansions.